File: verifiers/python/src/deviceintelligence_verifier/token_crypto.py
"""v2 ECIES token crypto (TokenCryptoV2.kt port) + the v1 keystream discriminator."""
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.serialization import NoEncryption
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
import hashlib
from . import keystream

PREFIX = "2:"
INFO_PREFIX = b"intel-token-v2"
HEADER = 1 + 1 + 32 + 12
TAG = 16


def is_v2(token: str) -> bool:
    """A v1 token is pure lowercase hex (no ':'); v2 carries the '2:' discriminator."""
    return token.startswith(PREFIX)


def decrypt(token_v2: str, server_scalar_or_priv) -> bytes:
    """Decrypt a '2:' token with the server X25519 private scalar.

    Raises InvalidTag on tamper, ValueError if malformed. Never returns plaintext
    on any tamper: every corruption fails the GCM tag.
    """
    if not token_v2.startswith(PREFIX):
        raise ValueError("not a v2 token")
    body = token_v2[len(PREFIX):]
    if len(body) % 2 != 0:
        raise ValueError("odd-length hex")
    try:
        p = bytes.fromhex(body)
    except ValueError:
        raise ValueError("bad hex char")
    if len(p) < HEADER + TAG:
        raise ValueError("v2 token too short")

    version, epoch = p[0], p[1]
    eph_pub, nonce = p[2:34], p[34:46]
    ct_and_tag = p[HEADER:]

    scalar = _scalar_of(server_scalar_or_priv)
    eph = bytearray(eph_pub)
    eph[31] &= 0x7F                       # RFC 7748: the ignored high bit
    shared = X25519PrivateKey.from_private_bytes(bytes(scalar)) \
        .exchange(X25519PublicKey.from_public_bytes(bytes(eph)))
    key = _hkdf_sha256(shared, nonce, INFO_PREFIX + bytes([epoch]), 32)

    aad = bytes([version, epoch]) + eph_pub
    return AESGCM(key).decrypt(nonce, ct_and_tag, aad)


def _scalar_of(key):
    from cryptography.hazmat.primitives.serialization import PrivateFormat
    from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey
    if isinstance(key, X25519PrivateKey):
        return key.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())
    return key.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())


def _hkdf_sha256(ikm: bytes, salt: bytes, info: bytes, length: int) -> bytes:
    return HKDF(algorithm=hashes.SHA256(), length=length, salt=salt,
                info=info).derive(ikm)

File: verifiers/node/src/token_crypto.ts
// v2 ECIES token crypto (TokenCryptoV2.kt port) + the v1 discriminator.
import { createDecipheriv, createHash, diffieHellman, generateKeyPairSync, createPublicKey } from "node:crypto";
import { hkdf } from "./hkdf.js";

const PREFIX = "2:";
const INFO_PREFIX = Buffer.from("intel-token-v2", "utf8");
const HEADER = 1 + 1 + 32 + 12;
const TAG = 16;

const SPKI_X25519_PREFIX = Buffer.from([
  0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x6e, 0x03, 0x21, 0x00]);

export function isV2(token: string): boolean {
  return token.startsWith(PREFIX);
}

function x25519PrivateFromRaw(scalar: Buffer) {
  const pkcs8 = Buffer.concat([Buffer.from("302e020100300506032b656e04220420", "hex"), scalar]);
  return createPrivateKey({ key: pkcs8, format: "der", type: "pkcs8" });
}

function x25519PublicFromRawLe(rawLe: Buffer) {
  const u = Buffer.from(rawLe);
  u[31] &= 0x7f; // RFC 7748: the ignored high bit
  return createPublicKey({
    key: Buffer.concat([SPKI_X25519_PREFIX, u]), format: "der", type: "spki",
  });
}

export function decryptV2(tokenV2: string, serverScalar: Buffer): Buffer {
  if (!tokenV2.startsWith(PREFIX)) throw new Error("not a v2 token");
  const body = tokenV2.slice(PREFIX.length);
  if (body.length % 2 !== 0) throw new Error("odd-length hex");
  let p: Buffer;
  try { p = Buffer.from(body, "hex"); } catch { throw new Error("bad hex char"); }
  if (p.length < HEADER + TAG) throw new Error("v2 token too short");

  const version = p[0], epoch = p[1];
  const ephPub = p.subarray(2, 34), nonce = p.subarray(34, 46);
  const ctAndTag = p.subarray(HEADER);

  const eph = Buffer.from(ephPub);
  eph[31] &= 0x7f;
  const shared = diffieHellman({
    privateKey: x25519PrivateFromRaw(serverScalar),
    publicKey: x25519PublicFromRawLe(eph),
  });
  const key = hkdf(shared, nonce, Buffer.concat([INFO_PREFIX, Buffer.of(epoch)]), 32);

  const aad = Buffer.concat([Buffer.of(version, epoch), ephPub]);
  const ct = ctAndTag.subarray(0, ctAndTag.length - TAG);
  const tag = ctAndTag.subarray(ctAndTag.length - TAG);
  const d = createDecipheriv("aes-256-gcm", key, nonce);
  d.setAAD(aad);
  d.setAuthTag(tag);
  return Buffer.concat([d.update(ct), d.final()]);
}

export { generateKeyPairSync, createHash };
